### ISSUE TITLE
importer: add Google Drive importer

### ISSUE THREAD

**Issue by bradfitz** [camlistore/camlistore#896]
It seems the Google Photos API (the "Picasaweb API") is broken and doesn't let you fetch all your photos. Details in #874.

But a workaround is to enable Google Drive's setting to mirror Google Photos in your Drive storage. (Go to https://drive.google.com/drive/my-drive then the Gear at the right, Settings > "Create a Google Photos folder" -> [X] Automatically put your Google Photos into a folder in My Drive)

Even if it weren't just for Google Photos, a Google Drive importer would be nice anyway, so let's do it.

/cc @mpl

**Comment by mpl** [comment]
Yep I've been wanting a drive importer anyway for a while too.

(on phone)

**Comment by tgulacsi** [comment]
The v3 api and the helps seems quite usable (https://developers.google.com/drive/v3/web/quickstart/go, https://godoc.org/google.golang.org/api/drive/v3).

If we use the drive.DrivePhotosReadonlyScope, then no other filtering is needed.
[list-files.zip](https://github.com/camlistore/camlistore/files/678510/list-files.zip)

**Comment by mpl** [comment]
For the importer to be useful for me (in itself, regardless of the google photos issue), I would need to be able to specify "directories", and/or files, to ignore. Does that seem like a reasonable feature to everyone?

Also, as opposed to our previous importers, some of the items in the drive (except for photos) are getting regularly updated, so how do we handle that? I mean, since we're content-addressable, it would be pretty easy for us to keep all versions of a file, but do we want to do that? I would prefer "overwriting" (i.e. dereferencing, so it eventually gets GCed) past versions of a file. Especially since drive has its own history/revert mechanism anyway. WDYAT?

**Comment by tgulacsi** [comment]
An alpha quality minimal version is available at https://camlistore-review.googlesource.com/9308
This works just as the Picasa importer, gets only the photos ("photos" space).

What kind of directories would you ignore (under the "photos" space)?

Iff all photos has been successfully imported, the changes are received through the "Changes" service, with the stored token, so receiving the changes are quite possible.

The dereferencing is new to me, so help wanted there...

**Comment by mpl** [comment]
@tgulacsi thanks, I'll try to have a look soon.

I don't think I'd need/want to ignore anything under the "photos space" as you say. I meant that if we're doing this, we might want to make it a generic gdrive importer anyway (not just for photos), in which case there are indeed files and directories that I'd like to ignore in my drive.

But maybe we only want photos for now if making a generic gdrive importer is much harder. we'll see.

**Comment by ianlewis** [comment]
Since Google Photos has a proper API now, is it maybe worth using that? https://developers.google.com/photos/

**Comment by mholt** [comment]
Just be aware that the new API strips a lot of the valuable metadata and exif data out of the photos and videos... If that's important then it's not yet a great solution.

**Comment by mpl** [comment]
@ianlewis excellent news, thanks. If this new API allows us to to actually get the latest version of the stored photo (as opposed to with the drive API, which is bugged that way), then it's definitely worth it.

**Comment by mpl** [comment]
@mholt Do you have an example or some documentation/report about that please?

**Comment by mpl** [comment]
Btw, none of these last comments change the fact that we also eventually want a drive importer for its own merit (not just for photos) anyway, which is kindof what this issue is about now.

For photos themselves, we should probably talk about it on https://github.com/perkeep/perkeep/issues/1144

**Comment by mholt** [comment]
@mpl Yes, I've been tracking it in https://github.com/mholt/photobak/issues/20.

Documentation: https://developers.google.com/photos/library/reference/rest/v1/mediaItems

Here's the bug report: https://issuetracker.google.com/issues/111228390

They did start including _some_ metadata for photos, but not videos. Here's the bug report for adding metadata for videos: https://issuetracker.google.com/issues/112010631

Still, it's probably better than nothing, unless you're already getting all the metadata using old APIs, in which case, switching to the new API would likely be a net loss.

**Comment by ivucica** [comment]
Agreed with @mpl, especially given the Photos<->Drive integration is seemingly going poof: https://github.com/perkeep/perkeep/issues/1144#issuecomment-502461294

**Comment by Masterxilo** [comment]
I would be interested in having this. However, what advantages over syncing from an rclone mounted Google Drive are we hoping for? Listening for changes? Full Google Drive v3 file model metadata? (the md5 attribute might come in handy, I have used it in the past to turn it into a  content adressable backend, however needs to be copied to custom prioerty for Google to index it for you...)
Conversion of .gdoc to .docx?
